docarray/doc_index/backends/hnswlib_doc_index.py

```diff
@@ -0,0 +1,400 @@
+        if kwargs:
+            raise ValueError(f'{list(kwargs.keys())} are not valid keyword arguments')
+        doc_seq = docs if isinstance(docs, Sequence) else [docs]
+        data_by_columns = self._get_col_value_dict(doc_seq)
+        hashed_ids = tuple(self._to_hashed_id(doc.id) for doc in doc_seq)
+
+        # indexing into HNSWLib and SQLite sequentially
+        # could be improved by processing in parallel
+        for col_name, index in self._hnsw_indices.items():
+            data = data_by_columns[col_name]
+            data_np = [self._to_numpy(arr) for arr in data]
+            data_stacked = np.stack(data_np)
+            index.add_items(data_stacked, ids=hashed_ids)
+            index.save_index(self._hnsw_locations[col_name])
+
+        self._send_docs_to_sqlite(doc_seq)
+        self._sqlite_conn.commit()
+
+    def execute_query(self, query: List[Tuple[str, Dict]], *args, **kwargs) -> Any:
+        if args or kwargs:
+            raise ValueError(
+                f'args and kwargs not supported for `execute_query` on {type(self)}'
+            )
+
+        ann_docs = DocumentArray.__class_getitem__(
+            cast(Type[BaseDocument], self._schema)
+        )([])
+        filter_conditions = []
+        doc_to_score: Dict[BaseDocument, Any] = {}
+        for op, op_kwargs in query:
+            if op == 'find':
+                docs, scores = self.find(**op_kwargs)
+                ann_docs.extend(docs)
+                doc_to_score.update(zip(docs.__getattribute__('id'), scores))
+            elif op == 'filter':
+                filter_conditions.append(op_kwargs['filter_query'])
+
+        docs_filtered = ann_docs
+        for cond in filter_conditions:
+            da_cls = DocumentArray.__class_getitem__(
+                cast(Type[BaseDocument], self._schema)
+            )
+            docs_filtered = da_cls(da_filter(docs_filtered, cond))
+
+        docs_and_scores = zip(
+            docs_filtered, (doc_to_score[doc.id] for doc in docs_filtered)
+        )
+        docs_sorted = sorted(docs_and_scores, key=lambda x: x[1])
+        out_docs, out_scores = zip(*docs_sorted)
+        return FindResult(documents=out_docs, scores=out_scores)
+
+    def _find_batched(
+        self,
+        query: np.ndarray,
+        search_field: str,
+        limit: int,
+        **kwargs,
+    ) -> FindResultBatched:
+        if kwargs:
+            raise ValueError(f'{list(kwargs.keys())} are not valid keyword arguments')
+        index = self._hnsw_indices[search_field]
+        labels, distances = index.knn_query(query, k=limit)
+        result_das = [
+            self._get_docs_sqlite_hashed_id(
+                ids_per_query.tolist(),
+            )
+            for ids_per_query in labels
+        ]
+        return FindResultBatched(documents=result_das, scores=distances)
+
+    @composable
+    def _find(
+        self, query: np.ndarray, search_field: str, limit: int, **kwargs
```

Comment:
as I said before, I think this one should accept `kwargs` but actually be specific

Reply:
After thinking about it, I don't think it works. If this doesn't take `**kwargs` but the public `find` does (as it needs to), then how does `find` pass the **kwargs to `_find`?
At runtime the **kwargs will be passed to a specific `_find` implementation that accepts all the needed kwargs, but as far as mypy is concerned, they are passed to the abstract `_find`.

So if that doesn't take `**kwargs` it will not work.

(Nor should it, from a typing perspective, bc we should not rely on the fact that a subclass implementation does something that the base class implementation doesn't do)